File: backend/report.py
from __future__ import annotations

from .models import DataTable, LabReport, NotBuiltYet


def markdown_table(data: DataTable) -> str:

    if not data.headers or not data.rows:
        return ""
    head = "| " + " | ".join(data.headers) + " |"
    sep = "| " + " | ".join("---" for _ in data.headers) + " |"
    body = []
    for row in data.rows:
        cells = list(row) + [""] * (len(data.headers) - len(row))
        body.append("| " + " | ".join(cells[: len(data.headers)]) + " |")
    return "\n".join([head, sep, *body])


def bullet_list(text: str) -> str:
    """Turn a multi-line string into Markdown bullets."""
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    return "\n".join(f"- {ln}" for ln in lines)


def numbered_list(text: str) -> str:
    """Turn a multi-line string into a numbered list"""
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    return "\n".join(f"{i}. {ln}" for i, ln in enumerate(lines, 1))


# ---------------------------------------------------------------------------
# The report itself
# ---------------------------------------------------------------------------


def build_report(report: LabReport) -> str:
    Info = report.info
    Content = report.content
    AnalysisQuestions = report.analysis
    Calcs = report.calculations
    table = []

    table.append(f"{Info.title or 'LabReport'}")
    table.append(
        f"**Name:** {Info.student_name} \n **Class:** {Info.course} \n **Date:** {Info.date} \n **Lab Partner/Partners:** {Info.partners}"
    )

    if Content.materials:
        table.append("# Material List\n\n" + bullet_list(Content.materials))

    if AnalysisQuestions:
        Questions = []
        for i, Q in enumerate(AnalysisQuestions, 1):
            Questions.append(f"**{i}. {Q.question}**\n\n{Q.answer}")
        table.append("## Analysis Questions\n\n" + "\n\n".join(Questions))

    if Calcs:
        calculatons = []
        for cal in Calcs:
            calculatons.append(f"**{cal.name} = {cal.pretty()}**\n\n {cal.work}")
        table.append("## Calculations\n\n" + "\n\n".join(calculatons))

    # One report can have several tables. Each prints under its own name.
    data_chunks = []
    for i, dt in enumerate(report.tables, 1):
        md = markdown_table(dt)
        if md:
            data_chunks.append(f"**{dt.title or f'Table {i}'}**\n\n{md}")
    if data_chunks:
        table.append("## Data\n\n" + "\n\n".join(data_chunks))

    return "\n\n".join(table)

File: tauri/src/backend/report.ts
/**
 * Builds the Markdown preview.  >>> THIS FILE IS YOURS TO WRITE. <<<
 *
 * One function: take a LabReport, return one Markdown string.
 *
 * THE RULES THAT BIT YOU LAST TIME, so they are written down now:
 *   - The lab title is the ONLY `#` in the whole document.
 *   - Every section below it is `##`. Not `#`, not `###`.
 *   - Section order is fixed:
 *         Material List -> Safety Precautions -> Data
 *         -> Calculations -> Analysis Questions
 *   - Safety is easy to forget. It goes in.
 *   - Any section with nothing in it is left out completely, not printed
 *     empty.
 */

import type { DataTable, LabReport } from "./models";
import { pretty } from "./models";

/** Turn a multi-line string into Markdown bullets. */
export function bulletList(text: string): string {
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => `- ${line}`)
    .join("\n");
}

/** Turn a multi-line string into a numbered list. */
export function numberedList(text: string): string {
  return text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line, i) => `${i + 1}. ${line}`)
    .join("\n");
}

/** One DataTable as a Markdown table. Returns "" if there's nothing in it. */
export function markdownTable(table: DataTable): string {
  if (table.headers.length === 0 || table.rows.length === 0) return "";
  const head = `| ${table.headers.join(" | ")} |`;
  const sep = `| ${table.headers.map(() => "---").join(" | ")} |`;
  const body = table.rows.map((row) => {
    const cells = [...row, ...Array(Math.max(table.headers.length - row.length, 0)).fill("")];
    return `| ${cells.slice(0, table.headers.length).join(" | ")} |`;
  });
  return [head, sep, ...body].join("\n");
}

/** The whole report as one Markdown string. */
export function buildReport(report: LabReport): string {
  const parts: string[] = [];
  parts.push(`# ${report.info.title || "Lab Report"}`);
  if (report.content.materials.trim()) {
    parts.push("## Material List\n\n" + bulletList(report.content.materials));
  }
  if (report.content.safety.trim()) {
    parts.push("## Safety Precautions\n\n" + bulletList(report.content.safety));
  }
  const tableChuncks: string[] = [];
  for (const table of report.tables) {
    const md = markdownTable(table);
    if (md === "") continue;
    if (table.title.trim()) {
      tableChuncks.push(`**${table.title}**\n\n${md}`);
    } else {
      tableChuncks.push(md);
    }
  }
  if (tableChuncks.length > 0) {
    parts.push("## Data\n\n" + tableChuncks.join("\n\n"));
  }
  const calcBlocks: string[] = [];
  for (const calc of report.calculations) {
    const lines = [`**${calc.name}**`, "", `- Formula: ${calc.formula}`];
    if (calc.work) lines.push(`- Work: ${calc.work}`);
    lines.push(`- Answer: **${pretty(calc)}**`);
    calcBlocks.push(lines.join("\n"));
  }
  if (calcBlocks.length > 0) {
    parts.push("## Calculations\n\n" + calcBlocks.join("\n\n"));
  }
  const asked = report.analysis.filter((qa) => qa.question.trim() || qa.answer.trim());
  if (asked.length > 0) {
    const blocks = asked.map((qa, i) => {
      const answer = qa.answer.trim() || "_(not answered)_";
      return `${i + 1}. **${qa.question.trim()}**\n\n   ${answer}`
    });
    parts.push("## Analysis Questions\n\n" + blocks.join("\n\n"));
  }

  return parts.join('\n\n');
}
